source/data-modeling/callbacks.txt

```diff
@@ -0,0 +1,173 @@
+.. _mongoid-modeling-callbacks:
+
+=========
+Callbacks
+=========
+
+.. facet::
+   :name: genre
+   :values: reference
+
+.. meta::
+   :keywords: ruby framework, odm, code example, life cycle
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+Overview
+--------
+
+In this guide, you can learn how to implement **callbacks** in your
+{+odm+} models to customize the life cycle of your model instances.
+
+Callbacks are methods that {+odm+} triggers at specified moments of
+an object's life cycle. They allow you to initiate specified actions
+before or after changes to an object's state.
+
+{+odm+} implements many of the callbacks from Active Record. To learn
+more, see `Callbacks
+<{+active-record-docs+}/active_record_callbacks.html>`__ in the
+Active Record documentation.
+
+Supported Callbacks
+-------------------
+
+Mongoid supports the following callbacks on model classes that implement
+the :ref:`Document <mongoid-modeling-documents>` module:
+
+-  ``after_initialize``
+-  ``after_build``
+-  ``before_validation``
+-  ``after_validation``
+-  ``before_create``
+-  ``around_create``
+-  ``after_create``
+-  ``after_find``
+-  ``before_update``
+-  ``around_update``
+-  ``after_update``
+-  ``before_upsert``
+-  ``around_upsert``
+-  ``after_upsert``
+-  ``before_save``
+-  ``around_save``
+-  ``after_save``
+-  ``before_destroy``
+-  ``around_destroy``
+-  ``after_destroy``
+
+To learn more about any of the preceding callback types, see the
+`ActiveRecord::Callbacks
+<https://api.rubyonrails.org/{+rails-8-version-docs+}/classes/ActiveRecord/Callbacks.html>`__
+reference in the Rails API documentation.
+
+You can implement a callback both top-level and embedded documents. 
+
+.. note:: Callback Invocation Behavior
+
+   For efficiency, {+odm+} invokes the callback only on the document
+   that you performed the persistence action on. This behavior enables
+   {+odm+} to support large hierarchies and handle optimized atomic
+   updates efficiently by not invoking callbacks throughout the document
+   hierarchy.
+
+Take precautions and ensure testability when implementing callbacks for
+domain logic, because these designs can lead to unexpected errors when
+callbacks in the chain halt execution. We recommend using callbacks for
+cross-cutting concerns, such as queueing up background jobs.
```

Comment:
Q: there's a couple terms in this paragraph (domain logic, cross-cutting) that I'm not familiar with - is it safe to assume readers should know these concepts, or do you think they're worth briefly explaining?

Reply:
I'll add some elaboration

Reply:
Domain logic should be a phrase users are familiar with